apps/provider.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/opensslconf.h>
+
+#include "apps.h"
+#include "app_params.h"
+#include "progs.h"
+#include <openssl/evp.h>
+#include <openssl/safestack.h>
+#include <openssl/provider.h>
+#include <openssl/core.h>
+#include <openssl/core_numbers.h>
+
+typedef enum OPTION_choice {
+    OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+    OPT_V = 100, OPT_VV, OPT_VVV
+} OPTION_CHOICE;
+
+const OPTIONS provider_options[] = {
+    {OPT_HELP_STR, 1, '-', "Usage: %s [options] provider...\n"},
+    {OPT_HELP_STR, 1, '-',
+        "  provider... Providers to load\n"},
+    {"help", OPT_HELP, '-', "Display this summary"},
+    {"v", OPT_V, '-', "List the algorithm names of specified provider"},
+    {"vv", OPT_VV, '-', "List the algorithm names of specified providers,"},
+    {OPT_MORE_STR, 0, '-', "categorised by operation type"},
+    {"vvv", OPT_VVV, '-', "List the algorithm names of specified provider"},
+    {OPT_MORE_STR, 0, '-', "one at a time, and list all known parameters"},
+    {NULL}
+};
+
+typedef struct info_st INFO;
+typedef struct meta_st META;
+
+struct info_st {
+    const char *name;
+    void *method;
+    const OSSL_PARAM *gettable_params;
+    const OSSL_PARAM *gettable_ctx_params;
+    const OSSL_PARAM *settable_ctx_params;
+};
+
+struct meta_st {
+    int first;                   /* For prints */
+    int total;
+    int indent;
+    int subindent;
+    int verbose;
+    const char *lable;
```

Comment:
NIT: label

Reply:
Argh...  "able" vs "label", I'll never get my head around that one...

Reply:
Fixed